### DISCUSSION TITLE
Running containerd on EKS 1.23

### DISCUSSION THREAD

**Discussion by jeffreymlewis** [gruntwork-io/knowledge-base#649] [How do I] [answered]
Hello,

I'm trying to use `containerd` on our EKS 1.23 cluster. The cluster is installed with `v0.100.0` of the Gruntwork service catalog [eks-cluster](https://github.com/gruntwork-io/terraform-aws-service-catalog/tree/master/modules/services/eks-cluster) module. I'm using Managed Node Groups. Unfortunately, adding `eks_bootstrap_script_options = "--container-runtime containerd"` to the node group configuration doesn't have the desired affect. I see the --container-runtime flag in the EC2 Launch Template, but the cloud-init script isn't using it. 🙁 

```
  managed_node_group_configurations = {
    "jlewis-${dependency.vpc.outputs.private_app_subnet_ids[0]}" = {
      subnet_ids                   = [dependency.vpc.outputs.private_app_subnet_ids[0]]
      eks_bootstrap_script_options = "--container-runtime containerd"
    }
  }
```

---

<ins datetime="2023-01-26T17:40:55Z">
  <p><a href="https://support.gruntwork.io/hc/requests/109841">Tracked in ticket #109841</a></p>
</ins>

**Comment by ryehowell** [comment, marked as answer]
Hi @jeffreymlewis,

How are you leveraging the `eks-cluster` module in your infrastructure code? For example, did you start with one of the [examples](https://github.com/gruntwork-io/terraform-aws-service-catalog/tree/master/examples) from the service-catalog, or did you write your own implementation/wrapper that leverages the modules in the service-catalog?

**Reply by jeffreymlewis** [reply]
HI @ryehowell 

We're using the Gruntwork Reference Architecture (is that still the name?) deployed by rho song in February 2021.

**Reply by ryehowell** [reply]
Hi @jeffreymlewis,

Can you confirm a few more details?
1. Is this a new EKS cluster, or is it an existing cluster that is being upgraded with pre-existing node groups? 
2. Can you confirm what AMI that is being used for the nodes?
3. Can you verify the runtime being used for the deployed nodes? ie `kubectl get nodes -o wide` and check the output of the `CONTAINER-RUNTIME` column

**Reply by jeffreymlewis** [reply]
Hi @ryehowell 

Thanks for your help with this! I'm adding a new Managed Node Group to a pre-existing cluster. (The old node group isn't changing, yet.)

All nodes in both Managed Node Group are on `docker://20.10.17`.

The launch template for the *new* MNG is using `ami-0eb3216fe26784e21`

Launch template for the *existing* MNGs are using `ami-0c9424a408e18bcc9`

**Reply by ryehowell** [reply]
@jeffreymlewis are the ami's specified custom built via the packer template provided in the service catalog or are they ami's built elsewhere?

**Reply by ryehowell** [reply]
Also, from the AWS Console, can you navigate to the Launch Template in question, look at the `Advanced Details` tab, and then look at the `User Data` section and look at the inline script? You should have something that looks like:
```
Content-Type: multipart/mixed; boundary="==BOUNDARY=="
MIME-Version: 1.0

--==BOUNDARY==
Content-Disposition: attachment; filename="_eks-worker-default-cloud-init"
Content-Transfer-Encoding: 7bit
Content-Type: text/x-shellscript
Mime-Version: 1.0

#!/bin/bash
#
# This script is meant to be run in the User Data of each EKS instance. This does the following:
#
# 1. Register the instance with the EKS cluster control plane.
# 1. Set node labels that map to the EC2 tags associated with the instance.
#
# Note, this script:
#
# 1. Assumes it is running in the AMI built from the eks-node-al2.json Packer template.
# 2. Has a number of variables filled in using Terraform interpolation.

set -e

...
# EXCLUDING BULK OF CONTENT. IMPORTANT PART IS TOWARD THE BOTTOM AND IS THE FOLLOWING:

bootstrap_args+=( --container-runtime containerd "$eks_cluster_name" )

echo "Running eks bootstrap script to register instance to cluster"
  /etc/eks/bootstrap.sh "${bootstrap_args[@]}"

...
--==BOUNDARY==--
``` 

Can you verify that line exists in the user data script?

**Reply by jeffreymlewis** [reply]
@ryehowell said,
> are the ami's specified custom built via the packer template provided in the service catalog or are they ami's built elsewhere?

We're not specifying a specific AMI using `cluster_instance_ami` nor `cluster_instance_ami_filters`. It looks like both default to `null`, which means the [image_id](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/launch_template#image_id) of the launch template is getting set to null. I guess this means AWS picks the latest AMI for eks 1.23?

**Reply by jeffreymlewis** [reply]
@ryehowell The user-data is as you showed above. In addition I see this at the bottom of the user-data script. Not sure if this means `bootstrap.sh` is being executed incorrectly?

```
/etc/eks/bootstrap.sh <redacted-cluster-name> --kubelet-extra-args '--node-labels=eks.amazonaws.com/sourceLaunchTemplateVersion=1,eks.amazonaws.com/nodegroup-image=ami-0eb3216fe26784e21,eks.amazonaws.com/capacityType=ON_DEMAND,eks.amazonaws.com/nodegroup=<redacted-cluster-name>-subnet-abcd12340,eks.amazonaws.com/sourceLaunchTemplateId=lt-abcd12340 --max-pods=58' --b64-cluster-ca $B64_CLUSTER_CA --apiserver-endpoint $API_SERVER_URL --dns-cluster-ip $K8S_CLUSTER_DNS_IP --use-max-pods false
```

**Reply by jeffreymlewis** [reply]
@ryehowell In `/var/log/cloud-init.log` I see cloud-init running the Gruntwork script (located at `/var/lib/cloud/instance/scripts/_eks-worker-default-cloud-init`) and it contains the snippet you showed above.

```
  bootstrap_args+=( --container-runtime containerd "$eks_cluster_name" )

  echo "Running eks bootstrap script to register instance to cluster"
  /etc/eks/bootstrap.sh "${bootstrap_args[@]}"
```

However, `/var/log/cloud-init-output.log` shows the following `bootstrap.sh` invocation, which tells me there's some kind of disconnect.

```
/etc/eks/bootstrap.sh <redacted-cluster-name> --kubelet-extra-args '--node-labels=eks.amazonaws.com/nodegroup-image=ami-0eb3216fe26784e21,eks.amazonaws.com/sourceLaunchTemplateVersion=2,eks.amazonaws.com/capacityType=ON_DEMAND,eks.amazonaws.com/nodegroup=<redacted-cluster-name>-subnet-3249732489327a787,eks.amazonaws.com/sourceLaunchTemplateId=lt-abcd1234 --max-pods=58' --b64-cluster-ca <redacted-cert> --apiserver-endpoint https://ABCD1234.gr7.us-east-1.eks.amazonaws.com --dns-cluster-ip 172.20.0.10 --use-max-pods false
```

**Reply by ryehowell** [reply]
Hi @jeffreymlewis, yes that looks like the issue and I just recreated the issue on my end. This issue arises when using the official EKS AMI's provided by AWS and not the hardened AMI's that are built via the packer template provided in [here](https://github.com/gruntwork-io/terraform-aws-service-catalog/blob/master/modules/services/eks-workers/eks-node-al2.pkr.hcl).

If you take a look at the [README](https://github.com/gruntwork-io/terraform-aws-service-catalog/tree/master/modules/services/eks-workers) for the `eks-workers` and the [docs site](https://docs.gruntwork.io/guides/build-it-yourself/kubernetes-cluster/deployment-walkthrough/pre-requisites), it goes thru the process for building and leveraging the hardened AMI's built by the GW module.

That being said, let me follow up shortly on possible workarounds to get those flags into the official AWS EKS AMI images if that is the route you would like to go. We do recommend leveraging the GW hardened AMI's instead of the AWS EKS AMI's though if at all possible. Give me a little bit to follow up.

**Reply by jeffreymlewis** [reply]
@ryehowell said,
> We do recommend leveraging the GW hardened AMI's ....

To be honest, I didn't know we could use custom AMIs with a MNG.

We originally switched to MNGs in an effort to simplify our deployment. If we're going to build custom AMIs anyway, is there an advantage to use MNGs over standard ASGs created with input variable `autoscaling_group_configurations` (in the `eks-cluster` module)?

**Reply by ryehowell** [reply]
HI @jeffreymlewis, that's a good question. MNG will simplify a *little* while providing some of the managed service capabilities that MNG offer (such as creating/managing the ASG(s)). Leveraging the SMNG will provide the greatest flexibility but with additional overhead. Since everything is already built via the IaC modules, most of this additional overhead is mostly transparent as it is already addressed in code. If looking for the least amount of overhead MNG would be the way to go, even with the AMI build as that would have to be done as well for the SMNG. 

MNG under the hood have always been LT's; only if not specified then AWS EKS automatically generates them. By providing custom LT's it just adds some additional flexibility to the end user. 

Building the AMI is pretty straightforward and handled via Packer. Only need to provide some input vars: 
- These are the general instructions: https://docs.gruntwork.io/guides/build-it-yourself/kubernetes-cluster/deployment-walkthrough/create-the-worker-node-ami
- Can use this Packer template (use the version of service catalog that coincides with your deployment): 
    - https://github.com/gruntwork-io/terraform-aws-service-catalog/blob/v0.100.0/modules/services/eks-workers/eks-node-al2.pkr.hcl
    - https://github.com/gruntwork-io/terraform-aws-service-catalog/tree/v0.100.0/modules/services/eks-workers

Here are a couple links that might be helpful:
A blog by Gruntwork explaining tradeoffs: https://blog.gruntwork.io/comprehensive-guide-to-eks-worker-nodes-94e241092cbe
AWS EKS References:
- https://docs.aws.amazon.com/eks/latest/userguide/eks-compute.html
- https://docs.aws.amazon.com/eks/latest/userguide/managed-node-groups.html
- https://docs.aws.amazon.com/eks/latest/userguide/launch-templates.html
- https://docs.aws.amazon.com/eks/latest/userguide/worker.html

**Reply by ryehowell** [reply]
Hi @jeffreymlewis, just to follow up regarding a workaround for passing in the `--container-runtime` flag to the official AWS EKS AMI (not a built AMI), this is possible by leveraging the `cloud_init_parts` config on the `managed_node_group_configurations` variable instead of using the `eks_bootstrap_script_options`.  Essentially you can pass in additional cloud-init config to get the additional flag needed to enable containerd on the AWS EKS AMI. 

Here is a working example:
```
managed_node_group_configurations = {
      # This is a containerd enabled MNG
      node_group = {
        min_size       = 1
        max_size       = 2
        subnet_ids     = [module.vpc.public_subnet_ids[0]]
        instance_types = ["t3.small"]
        tags           = var.managed_ec2_tags
        # eks_bootstrap_script_options = "--container-runtime containerd" # DON'T USE THIS
        # NOTE: this is based on the following issue: https://github.com/awslabs/amazon-eks-ami/issues/844
        cloud_init_parts = {
          first_part = {
            filename = "_eks-worker-boot-user-data"
            content_type = "text/x-shellscript"
            content = <<-EOT
              #!/bin/bash
              set -ex
              cat <<-EOF > /etc/profile.d/bootstrap.sh
              export CONTAINER_RUNTIME="containerd"
              EOF
              # Source extra environment variables in bootstrap script
              sed -i '/^set -o errexit/a\\nsource /etc/profile.d/bootstrap.sh' /etc/eks/bootstrap.sh
            EOT
          }
        }
      },
      # This is a default docker enabled MNG
      node_group_docker = {
        min_size       = 1
        max_size       = 2
        subnet_ids     = [module.vpc.public_subnet_ids[0]]
        instance_types = ["t3.small"]
        tags           = var.managed_ec2_tags
      }
    }
```

As mentioned in the comments, the `content` being provided is based on the following [Issue](https://github.com/awslabs/amazon-eks-ami/issues/844) in the amazon-eks-ami repo as getting the value set has been a bit of a headache for the broader community. 

Please let me know if this resolves the issue you were hitting and if there is anything else I can do to help 😄

**Reply by jeffreymlewis** [reply]
Thanks for the help @ryehowell . Given the convoluted nature of that work-around, it might actually be less complex to create a custom AMI (using a packer template).

Thanks for your help!

**Reply by ryehowell** [reply]
Glad I could help! Feel free to reach out anytime with any questions you may have. Happy coding!

**Reply by jeffreymlewis** [reply]
For others who might find this, note that it's not possible to change an EKS Managed Node Group from an "EKS Optimized AMI" to a custom AMI. Trying to do so using terraform, gives the following error.

```
error updating EKS Node Group (cluster-name:node-group-name) version: InvalidParameterException: You cannot specify an image id within the launch template, since your nodegroup is configured to use an EKS optimized AMI.
```

In this case, I had to delete the node group (`terragrunt destroy -target=...`) then run terragrunt apply to recreate it.

**Comment by jeffreymlewis** [comment]
Hi folks,

I see that [Tracked in ticket #109841](https://support.gruntwork.io/hc/requests/109841) was added to this KB article, but I'm not able to view this ticket on Zendesk. Can someone tell me the status of this ticket?

**Reply by ryehowell** [reply]
Hi @jeffreymlewis , 

We are looking into this today 😄. I actually think I see the issue, but need to validate it first. Hope to have a resolution for you shortly.
